### ISSUE TITLE
Fix Init State using hash mode

### ISSUE THREAD

**Issue by jorgecasar** [leeluolee/stateman#19]
Using `mode === HASH` the method `_fixInitState` do nothing. I think it should set the hash to `this.prefix + hashInPathName`.

I would like to suggest a refactor like following on `_fixInitState`: https://github.com/leeluolee/stateman/blob/master/src/histery.js#L179-L193

````javascript
_fixInitState: function(){
  var hash = this.location.hash.replace(this.prefix, "");
  if( this.mode === HISTORY && hash) {
    history.replaceState({}, document.title, _.cleanPath(this.root + hash));
  } else {
    var hashInPathName = _.cleanPath(this.location.pathname.replace(this.rRoot, ""));
    if(this.html5) {
      this.location.replace(this.root + this.prefix + hashInPathName);
    } else {
      this.location.hash = this.prefix + hashInPathName;
    }
  }
}
```
I made a work around using `history:change` event, but It's dirtier:

```javascript
stateman.on('history:change', function(path){
  if( path === '' ) {
    this.stateman.nav('/', {replace: true});
  }
}.bind(this))
```

**Comment by leeluolee** [comment]
In fact,  `mode` isn't a public option for developer .   The recommended way （see [Reference#start](http://leeluolee.github.io/stateman/?doc=API&lang=en#stateman-api-reference-api-statemanstart)） to pick `HASH` mode is: setting  `html5 !== true` or run stateman in browser that doesn't support `html5 history api` .  mode is only a private param for test (maybe it should prefixed with `_` , :(   ) . 

`_fixInitState` have two responsibility , and will work when starting stateman with flag `html5`.

1. the browser doesn't support `html5 history` and user visit site via `html5 way`, the code is 
 
 ```js
    if( this.mode !== HISTORY && this.html5) {
     if(hashInPathName)  // fix it
   }
  ```
  Then we can fix url from `/blog/list` to `/blog#/list`, If user visit `/blog/list` in __stupid & old__ browser . We think it highly possible that the url was pasted from other browser that support `html5` .  But we don't need to fix 
2.  As well as previous case, It is very probable that user visit site via __`Hash way` __ in modern browser that support html5.  the code: 

  ```js
   if( this.mode === HISTORY ) {
      if(hash) // fix it 
   }
  ```
  Then we can also fix It from `/blog#/list` to `/blog/list` . This case is also highly possible because user may use the url pasted from old browser.

Beyond that， `_fixInitState` do nothing else. 

So, your contribution may cause some issues, For example , visit the demo [`exmaple/layout.html`](https://github.com/jorgecasar/stateman/blob/fixInitState/example/layout.html)  . 

__But I find that `_fixInitState` is very obscure to understand Since It only works when `html5===true`.__

**Comment by leeluolee** [comment]
For `notfound` case. You can code like

```js

stateman.on('notfound', function(){
   this.go('app.homepage'); // or this.nav  as well
})
```
See [API](http://leeluolee.github.io/stateman/?doc=API&lang=en#stateman-api-reference-about-routing-routing-event) for more detail

`history:change` just a private event .

**Comment by jorgecasar** [comment]
The issue is that I'm using `html5=false` and first load of the page return a not found. I don't want the page 404, then redirection is not an option. First load of the app should redirect to /#/ o whatever the developer want to be the homepage insted of return a 404. If I have the state 404 and I redirect to this on notfound event, I have to check all the time if is or not the first time. I think this should be solved from the router. First time path is empty, I can check it but I think it's a extra job we can avoid if the config set a option to say what is the initial state of the application or set the diference between `location.pathname` and `confg.root`.

**Comment by leeluolee** [comment]
I think I know what you mean.

You have different logic between 『first load 』 and 『runtime 404』, right?

But, 

if using `html5=false`,   `root` and `_fixInitState` are all meaningless in this version.  

what you want is: 

- if user visit `/root/path/to` , fix it to `/root#/path/to`.  

But, you are using `html5=false`. It is almost impossible that user will use this url to visit your application.  

what you may need is :   

- if if user visit `/path` without hash , fix it to '#/'.  

I think it is a good idea. but instead of redirecting to `#/` automately,  we may need to provide a option to developer like

```js
stateman.start({
   defaults: '/user/id'
})

```

So, what is your opinion ?

**Comment by jorgecasar** [comment]
Yes, it's just I meant. I thought that `_fixInitState` was a good place to do that. Doing like I suggested, the migration to `html5 = true` would be less painful, in fact would be compatible. If you set `root: location.pathname` you have `defaults === '/'`.

What do you think about this solution?

```javascript
_fixInitState: function(){
    var pathname = _.cleanPath(this.location.pathname), hash, hashInPathName;
    // dont support history popstate
    if( this.mode !== HISTORY){
      hashInPathName = pathname.replace(this.rRoot, "")
      if( this.html5 ){
        // config the html5 mode
        if(hashInPathName) this.location.replace(this.root + this.prefix + hashInPathName);
      } else {
        // fallback to HASH mode.
        if( this.defaults ) {
          hashInPathName = this.defaults;
        }
        this.location.hash = this.prefix + hashInPathName;
      }
    }else if( this.mode === HISTORY /* && pathname === this.root*/){
      hash = this.location.hash.replace(this.prefix, "");
      if(hash) history.replaceState({}, document.title, _.cleanPath(this.root + hash))
    }
}
```
